src/python/pants/source/source_root.py

```diff
@@ -346,6 +346,8 @@ def fixed(self):
 
   def _do_add_pattern(self, pattern, langs, category):
     keys = pattern.split(os.path.sep)
+    if keys[-1] == '':
+      keys = keys[:-1]
```

Comment:
Changed this logic to explicitly check for normalization and raise an error if the path isn't equal to its normalization (just to be sure that the user is getting what they intend).

Also added some other validation checks, and quite a few more tests, including for a pattern at the root, as well as a fixed root at the root.